### PR TITLE
Add entity lifercycle interceptor interface and add a entity listener.

### PR DESCRIPTION
**Pull request by SirWayne**
Add entity lifercycle interceptor interface and add a entity listener which calls the interceptor callbacks

Signed-off-by: SirWayne <dennis.melzer@bosch-si.com>

### CODE REVIEW

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import javax.persistence.PostRemove;
+import javax.persistence.PostUpdate;
+import javax.persistence.PrePersist;
+import javax.persistence.PreRemove;
+import javax.persistence.PreUpdate;
+
+import org.eclipse.hawkbit.repository.jpa.model.helper.EntityInterceptorHolder;
+
+/**
+ * Entity listener which calls all entity interceptor for the lifecyles
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java]: Missing `s` => interceptors

Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java]: Maybe:  Entity listener which calls the callback's of all registered entity interceptors?

**Comment by michahirsch** [comment]
Tests are all failing because the application-context cannot be started.
You need to define the field `entityInterceptors` as optional.
```
@Autowired(required = false)
    private final List<EntityInterceptor> entityInterceptors = new ArrayList<>();
```

Otherwise it won't start because otherwise spring need at least one registered `EntityInterceptor`

**Comment by michahirsch** [comment]
It might make sense to add a test case in hawkBit which covers the functionality of calling the interceptor like this
```
/**
 * Copyright (c) 2015 Bosch Software Innovations GmbH and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package org.eclipse.hawkbit.repository.jpa.model;

import static org.fest.assertions.Assertions.assertThat;

import org.eclipse.hawkbit.repository.jpa.AbstractJpaIntegrationTest;
import org.eclipse.hawkbit.repository.jpa.model.helper.EntityInterceptorHolder;
import org.eclipse.hawkbit.repository.model.EntityInterceptor;
import org.eclipse.hawkbit.repository.model.Target;
import org.junit.Test;

import ru.yandex.qatools.allure.annotations.Features;
import ru.yandex.qatools.allure.annotations.Stories;

/**
 *
 */
@Features("Component Tests - Repository")
@Stories("Entity Listener Interceptor")
public class EntityInterceptorListenerTest extends AbstractJpaIntegrationTest {

    @Test
    public void postPersistIsCalledWhenPersistingATarget() {
        final MyEntityListener entityInterceptor = new MyEntityListener();
        EntityInterceptorHolder.getInstance().getEntityInterceptors().add(entityInterceptor);

        final Target targetToBeCreated = entityFactory.generateTarget("targetToBeCreated");

        targetManagement.createTarget(targetToBeCreated);

        assertThat(entityInterceptor.getPostPersistEntity()).isNotNull();
        assertThat(entityInterceptor.getPostPersistEntity()).isEqualTo(targetToBeCreated);
    }

    private static final class MyEntityListener implements EntityInterceptor {

        private Object postPersistEntity = null;

        @Override
        public void postPersist(final Object entity) {
            this.postPersistEntity = entity;
        }

        private Object getPostPersistEntity() {
            return postPersistEntity;
        }
    }
}

```

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * callbacks.
+ */
+public class EntityInterceptorListener {
+
+    @PrePersist
+    protected void prePersist(final Object entity) {
+        EntityInterceptorHolder.getInstance().getEntityInterceptors()
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java]: DRY - There is a shortcut  :wink:

`notifyAll(interceptor -> interceptor.prePersist(entity));`

```
private void notifyAll(final Consumer<? super EntityInterceptor> action) {
    EntityInterceptorHolder.getInstance().getEntityInterceptors().forEach(action);
}
```

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/helper/EntityInterceptorHolder.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /**
+     * @return the entity intreceptor holder singleton instance
+     */
+    public static EntityInterceptorHolder getInstance() {
+        return SINGLETON;
+    }
+
+    public List<EntityInterceptor> getEntityInterceptors() {
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/helper/EntityInterceptorHolder.java]: Be careful, you are returning your internal state! Everybody is able to modify it!

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import javax.persistence.PreRemove;
+import javax.persistence.PreUpdate;
+
+import org.eclipse.hawkbit.repository.jpa.model.helper.EntityInterceptorHolder;
+
+/**
+ * Entity listener which calls all entity interceptor for the lifecyles
+ * callbacks.
+ */
+public class EntityInterceptorListener {
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListener.java]: It isn't really a `Listener` -> `EntityInterceptor`?

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    @Test
+    @Description("Verfies that the post persist is called after a entity creation.")
+    public void postPersistIsCalledWhenPersistingATarget() {
+        executePersistAndAssertCallbackResult(new PostPersistEntityListener());
+    }
+
+    @Test
+    @Description("Verfies that the post load is called after a entity is loaded.")
+    public void postLoadIsCalledWhenLoadgATarget() {
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java]: Typo => `g` is wrong

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+        assertThat(entityInterceptor.getEntity()).isNotNull();
+        assertThat(entityInterceptor.getEntity()).isEqualTo(type);
+    }
+
+    private Target addListenerAndCreateTarget(final AbstractEntityListener entityInterceptor,
+            final Target targetToBeCreated) {
+        EntityInterceptorHolder.getInstance().getEntityInterceptors().add(entityInterceptor);
+
+        final Target createTarget = targetManagement.createTarget(targetToBeCreated);
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java]: return result immediately

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    @Test
+    @Description("Verfies that the post remove is called after a entity deletion.")
+    public void postRemoveIsCalledWhenDeletinggATarget() {
+        executeDeleteAndAssertCallbackResult(new PostRemoveEntityListener());
+    }
+
+    private void executePersistAndAssertCallbackResult(final AbstractEntityListener entityInterceptor) {
+        final Target targetToBeCreated = entityFactory.generateTarget("targetToBeCreated");
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java]: Why you're creating a target here? `addListenerAndCreateTarget(entityInterceptor, targetToBeCreated);` is doing this as well. It even returns it.

File: hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        addListenerAndCreateTarget(entityInterceptor, targetToBeCreated);
+
+        EntityInterceptorHolder.getInstance().getEntityInterceptors().remove(entityInterceptor);
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/test/java/org/eclipse/hawkbit/repository/jpa/model/EntityInterceptorListenerTest.java]: Why you remove the listener again?

**Pull request merged**